gax-grpc/src/main/java/com/google/api/gax/grpc/SafeShutdownManagedChannel.java

```diff
@@ -0,0 +1,150 @@
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.grpc;
+
+import io.grpc.CallOptions;
+import io.grpc.ClientCall;
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
+import io.grpc.ManagedChannel;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A {@link ManagedChannel} that will complete all calls started on the underlying channel before
+ * shutting down.
+ *
+ * <p>This class is not thread-safe. Caller must synchronize in order to ensure no new calls if safe
+ * shutdown has started.
+ *
+ * <p>Package-private for internal use.
+ */
+class SafeShutdownManagedChannel extends ManagedChannel {
+  private ManagedChannel delegate;
+  private AtomicInteger outstandingCalls = new AtomicInteger(0);
+  private volatile boolean isShutdownSafely = false;
+
+  SafeShutdownManagedChannel(ManagedChannel managedChannel) {
+    this.delegate = managedChannel;
+  }
+
+  /**
+   * Safely shutdown channel by checking that there are no more outstanding calls. If there are
+   * outstanding calls, the last call will invoke this method again when it complete
+   */
+  void shutdownSafely() {
+    isShutdownSafely = true;
+    if (outstandingCalls.get() == 0) {
+      delegate.shutdown();
+    }
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ManagedChannel shutdown() {
+    delegate.shutdown();
+    return this;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean isShutdown() {
+    return delegate.isShutdown();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean isTerminated() {
+    return delegate.isTerminated();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public ManagedChannel shutdownNow() {
+    delegate.shutdownNow();
+    return this;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    return delegate.awaitTermination(timeout, unit);
+  }
+
+  /**
+   * Wrap client call to decrement outstandingCalls and shutdown channel if necessary when it
+   * completes
+   *
+   * @param call call to be wrapped
+   * @return the wrapped call
+   */
+  private <ReqT, RespT> ClientCall<ReqT, RespT> clientCallWrapper(ClientCall<ReqT, RespT> call) {
+    return new SimpleForwardingClientCall<ReqT, RespT>(call) {
```

Comment:
Please extract this into an inner class, the doubly anonymous classes are a bit hard to read